psl-core/src/main/java/org/linqs/psl/reasoner/sgd/SGDReasoner.java

```diff
@@ -123,7 +123,10 @@ public double optimize(TermStore baseTermStore,
         // optimization pass because they are being updated in the variableUpdate() method.
         // Note that the number of variables may change in the first iteration (since grounding may happen then).
         double oldObjective = Double.POSITIVE_INFINITY;
-        float[] oldVariableValues = null;
+        float[] prevVariableValues = null;
+        // Save and use the variable values with the lowest computed objective
+        double lowestObjective = Double.POSITIVE_INFINITY;
+        float[] lowestVariableValues = null;
```

Comment:
Should the lowest objective semantics be handled through a config variable?
I legitimately do not know.
Is this something we would ever not want?

But more importantly, this doesn't seem like the right PR for this.

Reply:
Yeah, this should be a different PR.
It is pretty different than the online material, try to keep PRs to one topic.